velox/functions/sparksql/String.h

```diff
@@ -283,6 +286,78 @@ struct EndsWithFunction {
   }
 };
 
+/// Returns the substring from str before count occurrences of the delimiter
+/// delim. If count is positive, everything to the left of the final delimiter
+/// (counting from the left) is returned. If count is negative, everything to
+/// the right of the final delimiter (counting from the right) is returned. The
+/// function substring_index performs a case-sensitive match when searching for
+/// delim.
+template <typename T>
+struct SubstringIndexFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  bool isValidUTF8(const std::string& str) {
+    std::string utf8Str;
+    try {
+      utf8Str = boost::locale::conv::utf_to_utf<char>(str);
+    } catch (const boost::locale::conv::conversion_error&) {
+      return false;
+    }
+
+    return (utf8Str == str);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& str,
+      const arg_type<Varchar>& delim,
+      const int32_t& count) {
+    if (!(isValidUTF8(str) && isValidUTF8(delim))) {
+      result.setEmpty();
+      return;
+    }
+
+    if (count == 0) {
+      result.setEmpty();
+      return;
+    }
+
+    int64_t index;
+    if (count > 0) {
+      index = stringImpl::stringPosition<true, true>(str, delim, count);
+    } else {
+      index = stringImpl::stringPosition<true, false>(str, delim, -count);
+    }
+
+    auto start = 1;
+    auto length = stringImpl::length<true>(str);
+    auto delimLength = stringImpl::length<true>(delim);
+
+    /// Update the start and length based on the count value. And if the index
+    /// == 0, it will return the input str directly and no need to update the
+    /// start and length.
+    if (index != 0) {
+      if (count > 0) {
+        length = index - 1;
+      } else {
+        start = index + delimLength;
+        length = length - index - delimLength + 1;
+      }
+    }
+
+    auto byteRange = stringCore::getByteRange<true>(str.data(), start, length);
```

Comment:
Do we need to call stringCore::getByteRange? Can't we just create a StringView over a range of str.data()?

Reply:
No need. Updated.